src/api/StepsExchanger.ts

```diff
@@ -100,48 +99,81 @@ export class StepsExchanger extends CollabProvider {
     this.saveStatus = new ObservableString()
     this.debounce = saveWithDebounce()
     this.api = api
-    this.start()
     this.updateStoreVersion = updateStoreVersion
+
+    if (!navigator.onLine) {
+      this.saveStatus.setValue('offline')
+    } else {
+      this.start()
+    }
+
+    this.addNetworkListeners()
     this.stop = this.stop.bind(this)
   }
 
+  private addNetworkListeners() {
+    window.addEventListener('online', this.handleOnline)
+    window.addEventListener('offline', this.handleOffline)
+  }
+
+  private removeNetworkListeners() {
+    window.removeEventListener('online', this.handleOnline)
+    window.removeEventListener('offline', this.handleOffline)
+  }
+
+  private handleOnline = () => {
+    console.log('Connection recovered. Going back online.')
+    this.saveStatus.setValue('idle')
+    this.start()
+  }
+
+  private handleOffline = () => {
+    console.log('Connection lost. Going offline.')
+    this.saveStatus.setValue('offline')
+    this.closeConnection()
+  }
+
   async sendSteps(
     version: number,
     steps: Step[],
     clientID: string,
     flush = false
   ) {
-    if (steps.length === 0) {
-      return Promise.resolve()
-    }
 
     this.flushImmediately = this.debounce(
       async () => {
         this.saveStatus.setValue('saving')
 
-        const response = await this.api.sendSteps(
-          this.projectID,
-          this.manuscriptID,
-          {
+        const timeoutPromise = new Promise<{ error: string }>((resolve) => {
+          setTimeout(() => {
+            resolve({ error: 'timeout' })
+          }, REQUEST_TIMEOUT_MS)
+        })
+
+        const response = await Promise.race([
+          this.api.sendSteps(this.projectID, this.manuscriptID, {
             steps: steps,
             version,
             clientID,
-          }
-        )
+          }),
+          timeoutPromise,
```

Comment:
Why did you prefer to use Promise.race and not AbortController to timeout the response?